### PR TITLE
Enabling to share instance/history table

### PR DESCRIPTION
**Pull request by TsuyoshiUshio**
Hi @cgillum ,

As we discussed, I create a PR for DurableTask.  This PR will help someone who want to simplify Zero Downtime deployment for Durable Functions. 

 This PR enable us to configure separate ConnectionString between queue and Storage Table, also, increate AppId column to the Instance table with Query.

Please have a review of this PR. If it is accepted, I'll try to test via Durable.

### CODE REVIEW

**Comment by TsuyoshiUshio** [comment]
Hi @cgillum  Are you waiting for the PR of the Durable repo side? If so, I'll do it.

**Comment by cgillum** [comment]
Thank you for making this PR. I think the impact of this feature could be very big. not only does this allow side-by-side deployments, but it could also be the first step to enable hybrid deployments, something I've been thinking about recently. For example, maybe an orchestrator running in Azure could schedule an activity in another "app" which is hosted on premise.

No need to make a Durable Functions PR yet. Please give me a little more time to think about this one.

File: src/DurableTask.AzureStorage/AzureStorageOrchestrationService.cs [hunk 1]
```diff
@@ -1335,6 +1335,23 @@ public async Task<DurableStatusQueryResult> GetOrchestrationStateAsync(DateTime
...
+        /// </summary>
+        /// <param name="createdTimeFrom">CreatedTime of orchestrations. Fetch status grater than this value.</param>
+        /// <param name="createdTimeTo">CreatedTime of orchestrations. Fetch status less than this value.</param>
+        /// <param name="runtimeStatus">RuntimeStatus of orchestrations. You can specify several status.</param>
+        /// <param name="appIds">AppId of orchestration. You can specify several Ids.</param>
+        /// <param name="top">Top is number of records per one request.</param>
+        /// <param name="continuationToken">ContinuationToken of the pager.</param>
+        /// <param name="cancellationToken">Cancellation Token</param>
+        /// <returns>List of <see cref="OrchestrationState"/></returns>
+        public async Task<DurableStatusQueryResult> GetOrchestrationStateAsync(DateTime createdTimeFrom, DateTime? createdTimeTo, IEnumerable<OrchestrationStatus> runtimeStatus, string appIds, int top, string continuationToken, CancellationToken cancellationToken = default(CancellationToken))
```
Comment (cgillum) [inline src/DurableTask.AzureStorage/AzureStorageOrchestrationService.cs]: Two things:
1. The `appIds` should not be a string since it can be multiple parameters. Instead, it should be an `IEnumerable<string>` or something similar.
2. The list of parameters is getting too long. I think we should consider using a class to represent the filter. Maybe we can use your `OrchestrationInstanceStatusQueryCondition` class?

File: src/DurableTask.AzureStorage/Tracking/AzureTableTrackingStore.cs [hunk 2]
```diff
@@ -75,7 +75,8 @@ class AzureTableTrackingStore : TrackingStoreBase
             this.tableEntityConverter = new TableEntityConverter();
             this.taskHubName = settings.TaskHubName;
 
-            CloudStorageAccount account = CloudStorageAccount.Parse(settings.StorageConnectionString);
+            string storageConnectionString = settings.StorageTableConnectionString ?? settings.StorageConnectionString;
+            CloudStorageAccount account = CloudStorageAccount.Parse(storageConnectionString);
```
Comment (cgillum) [inline src/DurableTask.AzureStorage/Tracking/AzureTableTrackingStore.cs]: I think we should make the same change for the `InstanceStoreBackedTrackingStore` class.

File: src/DurableTask.AzureStorage/Tracking/TrackingStoreBase.cs [hunk 3]
```diff
@@ -69,6 +69,12 @@ public virtual Task<DurableStatusQueryResult> GetStateAsync(DateTime createdTime
             throw new NotSupportedException();
         }
 
+        /// <inheritdoc />
+        public virtual Task<DurableStatusQueryResult> GetStateAsync(DateTime createdTimeFrom, DateTime? createdTimeTo, IEnumerable<OrchestrationStatus> runtimeStatus, string appIds, int top, string continuationToken, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            throw new NotSupportedException();
```
Comment (cgillum) [inline src/DurableTask.AzureStorage/Tracking/TrackingStoreBase.cs]: Let's implement this for the other tracking store as well.

File: src/DurableTask.AzureStorage/AzureStorageOrchestrationServiceSettings.cs [hunk 4]
```diff
@@ -69,6 +69,17 @@ public class AzureStorageOrchestrationServiceSettings
...
+        /// <summary>
+        /// Gets or sets the Instance/History Table Storage connection string.
+        /// In case of null, StorageConnectionString is applied.
+        /// </summary>
+        public string StorageTableConnectionString { get; set; }
+
+        /// <summary>
+        /// Gets or sets the Application Id. 
+        /// </summary>
+        public string AppId { get; set; }
```
Comment (cgillum) [inline src/DurableTask.AzureStorage/AzureStorageOrchestrationServiceSettings.cs]: I'm not sure about the `AppId` name. I'm thinking about naming this `Zone` or something similar. The idea is that within a single task hub, you can define multiple zones which contain a distinct set of orchestrations and activities. Side-by-side deployments for versioning is one possible scenario. Maybe they can run in different hosting environments as well. I'm going to chat with Jeff about this idea.

Comment (TsuyoshiUshio) [reply src/DurableTask.AzureStorage/AzureStorageOrchestrationServiceSettings.cs]: Hi @cgillum  , I have a question about the concept of "Zone".  The fist idea was, just store the AppId (the name of FunctionApp) into the Instance table. Zone is a part of a single task hub. Then how they share the information with other Function Apps? And what is the example of the zone? then How we can retrieve the zone info since it can be multiple zones in a single task hub?

**Comment by TsuyoshiUshio** [comment]
Hi @cgillum , As we discussed, I remove all the new query and new attribute for instance table, and add these feature

* Separate Connection String for Instance/History table
* Create Zone and make it prefix of the History/Instance tables. If no Zone configured, it uses TaskHubName.

**Comment by TsuyoshiUshio** [comment]
@cgillum  I'll implement with AppId search again. If you think the AppId is proper, I'd happy to change it. (logic might be the same even if it change into TaskHub for example. )

File: src/DurableTask.AzureStorage/AzureStorageOrchestrationService.cs (hunk 1 above)
Comment (TsuyoshiUshio) [reply src/DurableTask.AzureStorage/AzureStorageOrchestrationService.cs]: I make it Condition object as you said.

File: src/DurableTask.AzureStorage/Tracking/AzureTableTrackingStore.cs (hunk 2 above)
Comment (TsuyoshiUshio) [reply src/DurableTask.AzureStorage/Tracking/AzureTableTrackingStore.cs]: InstanceStoreBackedTrakingStore has only this constructor. 

```
        public AzureTableInstanceStore(string hubName, string tableConnectionString)
        {
            this.tableClient = new AzureTableClient(hubName, tableConnectionString);

            // Workaround an issue with Storage that throws exceptions for any date < 1600 so DateTime.Min cannot be used
            DateTimeUtils.SetMinDateTimeForStorageEmulator();
        }
```

Also, the references are just Testing. I guess, Durable Functions use the Constructor, then it should be done by there.  Am I Correct? @cgillum

File: src/DurableTask.AzureStorage/Tracking/TrackingStoreBase.cs (hunk 3 above)
Comment (TsuyoshiUshio) [reply src/DurableTask.AzureStorage/Tracking/TrackingStoreBase.cs]: You mean move the AzureTableTrackingStore.GetStateAsync(OrchestrationInstanceStatusQueryCondition condition, int top, string continuationToken, CancellationToken cancellationToken = default(CancellationToken))
    to this place? It also include other private method might come here. Do I understand correctly?

File: src/DurableTask.AzureStorage/AzureStorageOrchestrationServiceSettings.cs (hunk 4 above)
Comment (TsuyoshiUshio) [reply src/DurableTask.AzureStorage/AzureStorageOrchestrationServiceSettings.cs]: I implement both Zone and AppId

**Comment by TsuyoshiUshio** [comment]
@cgillum I have only one questions. Also, if you don't like AppId, please let me know.

**Comment by TsuyoshiUshio** [comment]
Hi @cgillum, 

Could you review one more time? I rewrite whole.

**Comment by cgillum** [comment]
Yes, I will take a look now.

File: src/DurableTask.AzureStorage/AzureStorageOrchestrationServiceSettings.cs [hunk 5]
```diff
@@ -69,6 +69,24 @@ public class AzureStorageOrchestrationServiceSettings
         /// </summary>
         public string StorageConnectionString { get; set; }
 
+        /// <summary>
+        /// Gets or sets the Instance/History Table Storage connection string.
+        /// In case of null, StorageConnectionString is applied.
+        /// </summary>
+        public string StorageTableConnectionString { get; set; }
```
Comment (cgillum) [inline src/DurableTask.AzureStorage/AzureStorageOrchestrationServiceSettings.cs]: Let's rename this to `TrackingStoreConnectionString`.

**Comment by cgillum** [comment]
Actually, I will just put my main feedback here.

I've been thinking about this PR over the last several weeks because I think it is important for a lot of reasons. I want to simplify the design and get it merged soon.  Here are my current thoughts:

* I don't want to introduce a new "AppId" concept. Instead, I think we should use `TaskHub` as both the filter and the column name.
* I also don't want to introduce a new "Zone" concept. Instead, let's have a property which is just a prefix for the table names. We can call it `RemoteTrackingStoreNamePrefix`. This value is **only** used when `StorageTableConnectionString` is specified. The default value for `RemoteTrackingStoreNamePrefix` can be `DurableTask`. This means that by default we will create a table named `DurableTaskHistory` and `DurableTaskInstances`.

So basically, we don't introduce any new concepts - we just reuse existing concepts. The reason I want to do this is because I worry about creating confusion between concepts that exist for Azure Storage but not for other providers. I also don't want to speculate about other theoretical future use cases, since that can be dangerous and lead to technical debt.

Let me know what you think. I will also add some comments to the review.

File: src/DurableTask.AzureStorage/AzureStorageOrchestrationServiceSettings.cs (hunk 4 above)
Comment (cgillum) [reply src/DurableTask.AzureStorage/AzureStorageOrchestrationServiceSettings.cs]: Sorry, I changed my mind again :(.  I think there should be just one new property called `RemoteTrackingStoreNamePrefix`.

File: src/DurableTask.AzureStorage/Tracking/AzureTableTrackingStore.cs (hunk 2 above)
Comment (cgillum) [reply src/DurableTask.AzureStorage/Tracking/AzureTableTrackingStore.cs]: I see. If there are only test references, then we don't have to do anything here.

File: src/DurableTask.AzureStorage/Tracking/OrchestrationInstanceStatusQueryCondition.cs [hunk 6]
```diff
@@ -22,24 +22,45 @@ namespace DurableTask.AzureStorage.Tracking
...
 
+        /// <summary>
+        /// CreatedTimeTo
+        /// </summary>
         public DateTime CreatedTimeTo { get; set; }
 
+        /// <summary>
+        /// Collection of AppId
+        /// </summary>
+        public IEnumerable<string> AppIds { get; set; }
```
Comment (cgillum) [inline src/DurableTask.AzureStorage/Tracking/OrchestrationInstanceStatusQueryCondition.cs]: Per my other feedback, let's name this `TaskHubs`. I am hesitating about introducing new concepts.

File: src/DurableTask.AzureStorage/Tracking/TrackingStoreBase.cs (hunk 3 above)
Comment (cgillum) [reply src/DurableTask.AzureStorage/Tracking/TrackingStoreBase.cs]: Actually, let's not worry about it. This can be done later if/when somebody needs it.

**Comment by TsuyoshiUshio** [comment]
Hi @cgillum , 

This is the AppId discussion.

Imagine that we have production/staging slot for the deployment. 
If we use TaskHub concept, we need to change the taskhub name for every deployment. 
Also, we need to store somewhere which is which.  People might think they want to query for 
the target FunctionApp. 

I don't care if it is AppId or others. However, we might need something to query for the different FunctionApp. TaskHub is OK, but the operation might be complex. 

or Durable Functions Extension side pass the TaskHubName with FunctionApp name. Hmm... It seems not cool. If you have any comment, please let me know.

**Comment by cgillum** [comment]
I understand the concern. For the case that you mentioned, can customers use a combination of `%WEBSITE_SITE_NAME%` and `%WEBSITE_SLOT_NAME%` as the name of the task hub? That way it will change automatically.

In Durable 2.0, I want to make a breaking change where we derive the task hub name from the function app name by default. Let me know what you think.

**Comment by TsuyoshiUshio** [comment]
@cgillum It is good idea! It sounds cool! I'll start fix this PR.

File: src/DurableTask.AzureStorage/AzureStorageOrchestrationServiceSettings.cs (hunk 5 above)
Comment (cgillum) [reply src/DurableTask.AzureStorage/AzureStorageOrchestrationServiceSettings.cs]: Actually, I just remembered one thing. Someone recently contributed a new way of specifying the storage account: `StorageAccountDetails`. This allows people to use MSI instead of putting connection credentials directly into their configuration. Could you change your implementation to use `StorageAccountDetails` instead of `string` for the tracing store connection? Maybe then we would call it `TrackingStoreConnectionDetails`.

Comment (TsuyoshiUshio) [reply src/DurableTask.AzureStorage/AzureStorageOrchestrationServiceSettings.cs]: Done

File: src/DurableTask.AzureStorage/Tracking/OrchestrationInstanceStatusQueryCondition.cs (hunk 6 above)
Comment (TsuyoshiUshio) [reply src/DurableTask.AzureStorage/Tracking/OrchestrationInstanceStatusQueryCondition.cs]: Done

File: src/DurableTask.AzureStorage/AzureStorageOrchestrationServiceSettings.cs (hunk 4 above)
Comment (TsuyoshiUshio) [reply src/DurableTask.AzureStorage/AzureStorageOrchestrationServiceSettings.cs]: Done

File: src/DurableTask.AzureStorage/Tracking/AzureTableTrackingStore.cs (hunk 2 above)
Comment (TsuyoshiUshio) [reply src/DurableTask.AzureStorage/Tracking/AzureTableTrackingStore.cs]: Sure

File: src/DurableTask.AzureStorage/Tracking/OrchestrationInstanceStatusQueryCondition.cs (hunk 6 above)
Comment (TsuyoshiUshio) [reply src/DurableTask.AzureStorage/Tracking/OrchestrationInstanceStatusQueryCondition.cs]: However, I change it to TaskHubNames for consistency of the other parts.

**Comment by TsuyoshiUshio** [comment]
Hi @cgillum 

I've done it. However, something I need to tell you. 

1. I change the name of TaskHubs to TaskHubNames for the consistency
2. AzureTableTrackingStore seems change the constructor, it passes CloudStorageAccount object. I change the code because of the conflict. Test passes. Please review it, I'm ready to do more fix, If you have any comment, please let me know. I'll fix it today.

File: src/DurableTask.AzureStorage/Tracking/OrchestrationInstanceStatusQueryCondition.cs [hunk 7]
```diff
@@ -67,6 +88,16 @@ string GetConditions()
                 }
             }
 
+            if (this.TaskHubNames != null)
+            {
+                string appIdCondition = this.TaskHubNames.Select(x => TableQuery.GenerateFilterCondition("TaskHubName", QueryComparisons.Equal, x.ToString()))
```
Comment (cgillum) [inline src/DurableTask.AzureStorage/Tracking/OrchestrationInstanceStatusQueryCondition.cs]: I think this variable needs to be renamed to `taskHubCondition`.

File: src/DurableTask.AzureStorage/AzureStorageOrchestrationServiceSettings.cs [hunk 8]
```diff
@@ -72,6 +72,19 @@ public class AzureStorageOrchestrationServiceSettings
         /// </summary>
         public string StorageConnectionString { get; set; }
 
+        /// <summary>
+        /// Gets or sets the Instance/History Table Storage connection string.
+        /// In case of null, StorageConnectionString is applied.
+        /// </summary>
+        public string TrackingStoreConnectionString { get; set; }
```
Comment (cgillum) [inline src/DurableTask.AzureStorage/AzureStorageOrchestrationServiceSettings.cs]: I asked this in another comment already, but can we change this from `string` to `StorageAccountDetails`? Someone else introduced the `StorageAccountDetails` class, and it allows the user to configure with MSI.

File: src/DurableTask.AzureStorage/Tracking/AzureTableTrackingStore.cs [hunk 9]
```diff
@@ -75,16 +76,19 @@ class AzureTableTrackingStore : TrackingStoreBase
             this.stats = stats;
             this.tableEntityConverter = new TableEntityConverter();
             this.taskHubName = settings.TaskHubName;
+            this.remoteTrackingStoreNamePrefix = settings.RemoteTrackingStoreNamePrefix;
 
-            this.storageAccountName = account.Credentials.AccountName;
+            CloudStorageAccount trackingStoreAccount = string.IsNullOrEmpty(settings.TrackingStoreConnectionString) ? account : CloudStorageAccount.Parse(settings.TrackingStoreConnectionString);
```
Comment (cgillum) [inline src/DurableTask.AzureStorage/Tracking/AzureTableTrackingStore.cs]: Instead of changing this logic, can we change the caller of this constructor to pass in the correct `account` object? It doesn't make sense to pass an `account` parameter and then ignore it and create a new one.

File: src/DurableTask.AzureStorage/AzureStorageOrchestrationServiceSettings.cs (hunk 8 above)
Comment (cgillum) [reply src/DurableTask.AzureStorage/AzureStorageOrchestrationServiceSettings.cs]: See this PR for more details: https://github.com/Azure/durabletask/pull/248

Comment (TsuyoshiUshio) [reply src/DurableTask.AzureStorage/AzureStorageOrchestrationServiceSettings.cs]: Done

File: src/DurableTask.AzureStorage/Tracking/AzureTableTrackingStore.cs (hunk 9 above)
Comment (TsuyoshiUshio) [reply src/DurableTask.AzureStorage/Tracking/AzureTableTrackingStore.cs]: Done

**Comment by TsuyoshiUshio** [comment]
Hi @cgillum, 

I've done all fix. 

According to your comment, I change TrakingStoreConnectionString from string to StorageAccountDetails. 

However, things is not simple for this time. As you commented on the #248 , you want to obsolete the StorageConnectionString. We need to add the StorageConnectionString to the StorageAccountDetails with Automatically decide if it is StorageConnectionString or Creadentials.

Also, according to the GRASP pattern, I simplify the AzureStorageOrchestrationServiceSettings. 

If you don't like this design, I'd happy to refactor that with the highest priority or you can fix by your self. if you have no time.

File: Test/DurableTask.AzureStorage.Tests/AzureStorageOrchestrationServiceSettingsTest.cs [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+            Assert.AreEqual("fooInstances", settings.InstanceTableName);
+        }
+
+        [TestMethod]
+        public void TrackingStoreStorageAccountDetailsHasSet()
+        {
+            var settings = new AzureStorageOrchestrationServiceSettings()
+            {
+                TaskHubName = "foo",
+                RemoteTrackingStoreNamePrefix = "bar",
```
Comment (cgillum) [inline Test/DurableTask.AzureStorage.Tests/AzureStorageOrchestrationServiceSettingsTest.cs]: I sincerely apologize, but I changed my mind a bit. I realized that I don't like the "Remote" part of this property name. Instead, I think it should just be `TrackingStoreNamePrefix` because "Remote" doesn't really mean anything in this context.

I'm really glad you wrote unit tests to make sure this behaves correctly. Thank you!

File: src/DurableTask.AzureStorage/StorageAccountDetails.cs [hunk 11]
```diff
@@ -34,5 +34,10 @@ public sealed class StorageAccountDetails
         /// The storage account endpoint suffix
         /// </summary>
         public string EndpointSuffix { get; set; }
+
+        /// <summary>
+        /// The storage account connection string
+        /// </summary>
+        public string StorageAccountConnectionString { get; set; }
```
[the remark this answers is not in the document]
Comment (TsuyoshiUshio) [reply src/DurableTask.AzureStorage/StorageAccountDetails.cs]: Hi @cgillum , 

If so, we need TrakingStoreStorageAccountConnectionString.  I read your issue and I thought you want to put something in the StorageAccountDetails. That is why I add it here. Also I read you want to obsolete the StorageAccountConnectionString in the near future.  How can we set the ConnectionString for StorageAccountDetails? Don't need it?

[the remark this answers is not in the document]
Comment (TsuyoshiUshio) [reply src/DurableTask.AzureStorage/StorageAccountDetails.cs]: Self resolved. StorageCredentials class has a method of tryParseCredentials for ConnectionString. Sorry for that.

[the remark this answers is not in the document]
Comment (TsuyoshiUshio) [reply src/DurableTask.AzureStorage/StorageAccountDetails.cs]: Removed

File: Test/DurableTask.AzureStorage.Tests/AzureStorageOrchestrationServiceSettingsTest.cs (hunk 10 above)
Comment (TsuyoshiUshio) [reply Test/DurableTask.AzureStorage.Tests/AzureStorageOrchestrationServiceSettingsTest.cs]: Done

File: src/DurableTask.AzureStorage/StorageAccountDetailsExtensions.cs [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Text;
+    using Microsoft.WindowsAzure.Storage;
+
+    /// <summary>
+    /// Extension methods for <see cref="StorageAccountDetails"/>
+    /// </summary>
+    public static class StorageAccountDetailsExtensions
```
[the remark this answers is not in the document]
Comment (TsuyoshiUshio) [reply src/DurableTask.AzureStorage/StorageAccountDetailsExtensions.cs]: Remove this. Make it instance method of StorageAccountDetails.

File: src/DurableTask.AzureStorage/StorageAccountDetails.cs [hunk 13]
```diff
@@ -36,8 +37,8 @@ public sealed class StorageAccountDetails
         public string EndpointSuffix { get; set; }
 
         /// <summary>
-        /// The storage account connection string
+        ///  CloudStorageAccount 
         /// </summary>
-        public string StorageAccountConnectionString { get; set; }
+        public CloudStorageAccount CloudStorageAccount => new CloudStorageAccount(this.StorageCredentials, this.AccountName, this.EndpointSuffix, true);
```
Comment (cgillum) [inline src/DurableTask.AzureStorage/StorageAccountDetails.cs]: Last thing: Generally speaking, .NET properties should not allocate like this. Customers will not expect this kind of behavior and it can lead to unexpected problems. Instead, make it a method, like `ToCloudStorageAccount()`.

**Comment by TsuyoshiUshio** [comment]
Hi @cgillum 
I finish the change. However, I encounter the issue of testing. MultiWorkerLeaseMovement fails. However, there is no reason this one fails. I makes sure it works related Connection String. Could you run the test on your environment? I'll keep on investigate it.

File: src/DurableTask.AzureStorage/StorageAccountDetails.cs (hunk 13 above)
Comment (TsuyoshiUshio) [reply src/DurableTask.AzureStorage/StorageAccountDetails.cs]: Done
